src/crimson/os/seastore/journal.h

```diff
@@ -26,13 +26,101 @@ namespace crimson::os::seastore {
 class SegmentProvider;
 class SegmentedAllocator;
 
+class Journal {
+public:
+  using open_for_write_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error
+    >;
+  using open_for_write_ret = open_for_write_ertr::future<journal_seq_t>;
+  virtual open_for_write_ret open_for_write() = 0;
+
+  using close_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  virtual close_ertr::future<> close() = 0;
+
+  using submit_record_ertr = crimson::errorator<
+    crimson::ct_error::erange,
+    crimson::ct_error::input_output_error
+    >;
+  using submit_record_ret = submit_record_ertr::future<
+    record_locator_t
+    >;
+  virtual submit_record_ret submit_record(
+    record_t &&record,
+    OrderingHandle &handle
+  ) = 0;
+
+  virtual seastar::future<> flush(OrderingHandle &handle) = 0;
+
+  virtual void set_write_pipeline(WritePipeline *_write_pipeline) = 0;
+  virtual ~Journal() {}
+};
+using JournalRef = std::unique_ptr<Journal>;
+
```

Comment:
I don't understand the purpose of JournalManager.  What is the purpose of having more than one journal?

Reply:
If I understand segment manager and RBM design correctly, Seastore needs to handle multiple devices (segmented and rbm) ultimately. So, the purpose of this JournalManager is to forward the record to a target journal (either segmented journal or circularboundedjournal) depending on the target location. Did I miss something?

I think JournalManager can be removed for the time being.

Reply:
Multiple devices, yes.  However, the purpose of a Journal is to persistently enforce ordering after a crash.  Writing different records to different journals would break those ordering guarantees (at least, without additional complications this PR doesn't seem to have).  The ExtentPlacementManager is how we're writing to devices other than the journal device, but for now there should only be a single Journal instance for any single SeaStore instance.  The existing implementation writes to a SegmentManager.  We need another implementation to allow us to journal to a RBM device instead, when one exists, so we need to generalize the existing interface to allow the users to transparently use whichever of the two implementations happens to exist.  We want to avoid other components depending specifically on SegmentJournal.

Reply:
In a configuration with both a fast nvme device and a slower zns device, I'd expect that we'd have a single CBJournal on the faster nvme device with extents being split between an RBM device on the remainder of the fast nvme device and a SegmentManager on the slower ZNS device.  Does that seem reasonable?

Reply:
If I understand correctly, you mean the right way is CBJournalAllocator inherited from ExtentAllocator in the ExtentPlacementManager, not the Journal (proposed this PR)? 

If so, how we can submit the record using EPM in submit_transaction_direct in case of CBJournal. something like epm->submit_record(), not journal->submit_record()?

> In a configuration with both a fast nvme device and a slower zns device, I'd expect that we'd have a single CBJournal on the faster nvme device with extents being split between an RBM device on the remainder of the fast nvme device and a SegmentManager on the slower ZNS device. Does that seem reasonable?

Yes.

Reply:
I'm not sure I understand what you mean.  I'd have expected that we could simply change Journal to an interface with the existing implementation renamed to SegmentedJournal inheriting from Journal.  We'd then add another implementation, CBJournal, also inheriting from Journal.  On startup, either a CBJournal or a SegmentedJournal would be created and passed to all components as Journal.

Reply:
I think that is what I intended for this PR except for JournalManager.  If JournalManager is removed in this PR,  does these commits already address your comments (even though the commits needs to be improved)? ExtentPlacementManager can take over a role of JournalManager later.

Reply:
Replacing JournalManager dependencies with with Journal should work. TransactionManager should depend on Journal rather than JournalManager as you've said.  EPM should depend on the general Journal interface rather than SegmentJournal.  

We won't need EPM to replace the JournalManager.  Writing to devices other than the journal device doesn't mean writing to another journal, it means doing an out-of-line (ool) allocation on another device prior to the journal commit.  That capability is partly already in place, see ExtentPlacementManager::delayed_alloc_or_ool_write and related code and @cyx1231st's pending PR https://github.com/ceph/ceph/pull/44809.  Writing to an RBM device rather than than a SegmentManager device would mean extending the existing ExtentAllocator interface to allocate from an RBM device as the SegmentedAllocator does from a SegmentedDevice.

My current thinking is that we probably won't ever need to permit more than one journal device.  If we ever choose to permit multiple journal devices, we'll do so under the Journal interface.

I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space.  Existing unit tests where appropriate should be updated to also test that configuration.  The step after that would be to integrate an RBM allocator to do ool writes to RBM devices.

Does that make sense?

Reply:
> I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space.

IIUC this is referring to the TODOs about:
1. https://github.com/ceph/ceph/blob/32da0e0ecb70224d005deb13bda8808f2665a811/src/crimson/os/seastore/extent_placement_manager.h#L313
2. https://github.com/ceph/ceph/blob/32da0e0ecb70224d005deb13bda8808f2665a811/src/crimson/os/seastore/extent_placement_manager.h#L318

Forcing all extents to be written ool means RBM will be able to decouple all extents out of journal records (i.e. inlining), so that extents can be decoupled out of journaling & trimming processes with their allocations decided by the RBM allocator. And journals can be bounded to a reserved space effectively if without inline extents. Hopefully with this direction we could reuse the existing Journal logic as much as possible.

This will also benefit segmented manager as we will be able to classify cold and hot extents more effectively towards better segmented GC.

BTW I'm working on 2 with further refactors to generalize EPM.

Does this also make sense to you @athanatos ?

Reply:
> We won't need EPM to replace the JournalManager. Writing to devices other than the journal device doesn't mean writing to another journal, it means doing an out-of-line (ool) allocation on another device prior to the journal commit. That capability is partly already in place, see ExtentPlacementManager::delayed_alloc_or_ool_write and related code and @cyx1231st's pending PR https://github.com/ceph/ceph/pull/44809. Writing to an RBM device rather than than a SegmentManager device would mean extending the existing ExtentAllocator interface to allocate from an RBM device as the SegmentedAllocator does from a SegmentedDevice.

I got it.

> My current thinking is that we probably won't ever need to permit more than one journal device. If we ever choose to permit multiple journal devices, we'll do so under the Journal interface.

I agree.

> I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space. Existing unit tests where appropriate should be updated to also test that configuration. The step after that would be to integrate an RBM allocator to do ool writes to RBM devices.

Seems reasonable to me.

Reply:
@cyx1231st Yeah, I think that's mostly right.